embassy-time/src/queue.rs

```diff
@@ -0,0 +1,617 @@
+        self.alarm_at = Instant::MAX;
+
+        self.dispatch(alarm_schedule);
+    }
+}
+
+/// The generic queue implementation
+pub struct Queue<const N: usize = 128, R: RawMutex = CriticalSectionRawMutex> {
+    inner: Mutex<R, RefCell<InnerQueue<N>>>,
+    alarm: Cell<Option<AlarmHandle>>,
+    alarm_schedule: AtomicU64,
+}
+
+impl<const N: usize, R: RawMutex + 'static> Queue<N, R> {
+    /// Create a Queue
+    pub const fn new() -> Self {
+        Self {
+            inner: Mutex::new(RefCell::new(InnerQueue::<N>::new())),
+            alarm: Cell::new(None),
+            alarm_schedule: AtomicU64::new(u64::MAX),
+        }
+    }
+
+    /// Initialize the queue
+    ///
+    /// This method is called from [`initialize`](crate::queue::initialize), so you are not expected to call it directly.
+    /// Call [`initialize`](crate::queue::initialize) instead.
+    ///
+    /// # Safety
+    /// It is UB call this function more than once, or to call it after any of your
+    /// futures that use `embassy-time` are polled already.
+    pub unsafe fn initialize(&'static self) {
+        if self.alarm.get().is_some() {
+            panic!("Queue is already initialized");
+        }
+
+        let handle = allocate_alarm().unwrap();
+        self.alarm.set(Some(handle));
+
+        set_alarm_callback(handle, Self::handle_alarm_callback, self as *const _ as _);
+    }
+
+    /// Schedule a new waker to be awoken at moment `at`
+    ///
+    /// This method is called internally by [`embassy-time`](crate), so you are not expected to call it directly.
+    pub fn schedule(&'static self, at: Instant, waker: &Waker) {
+        self.check_initialized();
+
+        self.inner
+            .lock(|inner| inner.borrow_mut().schedule(at, waker, &self.alarm_schedule));
+
+        self.update_alarm();
+    }
+
+    fn check_initialized(&self) {
+        if self.alarm.get().is_none() {
+            panic!("Queue is not initialized yet");
+        }
+    }
+
+    fn update_alarm(&self) {
+        // Need to set the alarm when we are *not* holding the mutex on the inner queue
+        // because mutexes are not re-entrant, which is a problem because `set_alarm` might immediately
+        // call us back if the timestamp is in the past.
+        let alarm_at = self.alarm_schedule.swap(u64::MAX, AtomicOrdering::SeqCst);
+
+        if alarm_at < u64::MAX {
+            set_alarm(self.alarm.get().unwrap(), alarm_at);
+        }
+    }
+
+    fn handle_alarm(&self) {
+        self.check_initialized();
+        self.inner
+            .lock(|inner| inner.borrow_mut().handle_alarm(&self.alarm_schedule));
+
+        self.update_alarm();
+    }
+
+    fn handle_alarm_callback(ctx: *mut ()) {
+        unsafe { (ctx as *const Self).as_ref().unwrap() }.handle_alarm();
+    }
+}
+
+unsafe impl<const N: usize, R: RawMutex + 'static> Send for Queue<N, R> {}
+unsafe impl<const N: usize, R: RawMutex + 'static> Sync for Queue<N, R> {}
```

Comment:
This is not right because `Cell` is not Sync. This can be a problem in practice if the user calls `.initialize()` from one thread and `.update_alarm()` from another (through polling a timer), which would cause a data race on the `alarm` Cell.

Reply:
That's precisely the reason why `Queue::initialize` is unsafe - because it has to be called *before* any futures that use `embassy-time` are polled.

I can surely protect the `Cell` instance with a mutex as well, but this mutex - at least in the RTOS case - might need to be of a different type than the one protecting the inner queue structure, I'm afraid. I'll elaborate on that further down, where the `initialize` method itself is being discussed.